### PR TITLE
Move guard optimizations from v3_kernel to a new beam_ssa_bool pass

### PR DESCRIPTION
**Pull request by bjorng**
In 348b5e6bee2f, I added a guard optimization pass to `v3_kernel`. At the time, Kernel Erlang was the intermediate representation most suited for doing that kind of optimization. However, it was not ideal from a maintenance standpoint, because none of the current compiler maintainers have a deep grasp of Kernel Erlang. The existence of the optimizations could also make other improvements more difficult (such as generating SSA code directly from Core Erlang).

Therefore, since we now have SSA code, it is time to retire the guard optimizations in v3_kernel and rewrite the optimizations to work on SSA code. These optimizations gives roughly as good as result as the old ones.

See the beginning of the `beam_ssa_bool.erl` file for more detailed information about the optimizations.

### CODE REVIEW

File: lib/compiler/src/beam_ssa_bool.erl [hunk 1]
```diff
@@ -0,0 +1,1680 @@
...
+        throw:true ->
+            true
+    end.
+
+dg_reverse_postorder(G, Vs) ->
+    Seen = cerl_sets:new(),
+    {RPO,_} = dg__rpo(Vs, G, Seen, []),
+    RPO.
+
+dg__rpo([V|Vs], G, Seen0, Acc0) ->
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/beam_ssa_bool.erl]: Yes, I use them for helper functions of the digraph functions.

File: lib/compiler/src/beam_ssa_bool.erl [hunk 2]
```diff
@@ -0,0 +1,1680 @@
...
+
+%%%
+%%% Optimize the graph, attempting to eliminating 'and', 'or', and 'not'
+%%% instructions.
+%%%
+
+opt_digraph_top(Arg, G0, St) ->
+    #b_set{dst=Dst} = I = get_def(Arg, G0, St),
+    {br,Succ,Fail} = get_targets(Dst, G0, St),
+    #b_set{op={bif,'=:='},dst=Dst,
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/beam_ssa_bool.erl]: Thanks, fixed!

**Comment by michalmuskala** [comment]
Looking at this, I wonder if it would be possible to make this optimisation more efficient on boolean code outside of guards. In particular there's a pattern of how Elixir compiles `if` that would be awesome if it could be optimised:
```erlang
case Expr of
  Tmp when Tmp =:= false; Tmp =:= nil -> %% false branch
  _ -> %% true branch
end
```
When `Expr` is purely a boolean expression, this should be handled by this pass, but unfortunately the type strength reduction pass runs after this one, so the comparison with `nil` won't be dropped by the time this runs. Furthermore, this only handles comparison with `true` and not with `false` - it might be a useful transformation for the preopt pass to switch those around if possible.

**Comment by bjorng** [comment]
`when Tmp =:= false; Tmp =:= nil` **is** a guard, and `beam_ssa_bool` can optimize it. That is, it will get rid of the call to the `or` BIF. If `Tmp` can later be proven to be a boolean, `beam_ssa_type` will get rid of the comparison with `nil`.

I compiled this code:

```
foo() ->
    Expr = get_some_value(),
    case Expr of
        Tmp when Tmp =:= false; Tmp =:= nil -> error;
        _ -> ok
    end.

bar(A, B) ->
    Expr = A > B,
    case Expr of
        Tmp when Tmp =:= false; Tmp =:= nil -> error;
        _ -> ok
    end.

get_some_value() ->
    get(value).
```
`beam_ssa_bool` will get rid of the `or` BIF in both `foo/0` and `bar/2`. (Use the `dbool` option to produce a `.bool` file with the output from `beam_ssa_bool`.)

`beam_ssa_type` will get rid of the comparison with `nil` in `bar/2`, but not in `foo/0`.

That said, if you find any examples where the compiler gives a worse result than in OTP 22, I am of course interested.

**Pull request merged**